core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,166 @@
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    /// @param origin The address which initiated the lock.
+    /// @param recipient The substrate address that will receive the funds.
+    /// @param token The token locked.
+    /// @param amount The amount locked.
+    event Locked(address origin, bytes32 recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    /// @param origin The substrate address which initiated the unlock.
+    /// @param recipient The ethereyn address that will receive the funds.
+    /// @param token The token unlocked.
+    /// @param amount The amount unlocked.
+    event Unlocked(bytes32 origin, address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    /// @param token The address of the token created.
+    // TODO: Remove name, symbol and decimals.
+    event Created(address token, string name, string symbol, uint8 decimals);
+
+    /// @dev Zero amount.
+    error ZeroAmount();
+
+    /// @dev Zero addressed token.
+    error ZeroAddressToken();
+
+    /// @dev Zero addressed recipient.
+    error ZeroAddressRecipient();
+
+    /// @dev the origin provided is unauthorized.
+    error UnauthorizedOrigin();
+
+    /// @dev the message action is unsupported.
+    error UnsupportedMessageAction();
+
+    /// @dev the origin that
+    bytes32 public immutable allowedOrigin;
+
+    /// @dev The vault where ERC20 tokens are locked.
+    ERC20Vault public immutable vault;
+
+    /// @dev The channel used to enqueue messages for lock and create functions.
+    OutboundChannel public immutable outboundChannel;
+
+    /// @dev Initializes the NativeTokens contract with a vault and channels.
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    /// @param _allowedOrigin The origin allowed to call handle.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel, bytes32 _allowedOrigin) {
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        allowedOrigin = _allowedOrigin;
+    }
+
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message. Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes32 recipient, uint128 amount) public {
+        if (amount == 0) {
+            revert ZeroAmount();
+        }
+        if (token == address(0)) {
+            revert ZeroAddressToken();
+        }
+        if (recipient == bytes32(0)) {
+            revert ZeroAddressRecipient();
+        }
```

Comment:
These checks are unnecessary, I think. If the token is 0, then the transfer functions will revert. If the recipient is zero, the user definitely provided invalid input and their funds will be lost.

Reply:
addressed in 77bd470fe95a39.